lang/src/org/partiql/lang/eval/visitors/OrderByAliasVisitorTransform.kt

```diff
@@ -0,0 +1,81 @@
+import org.partiql.lang.domains.PartiqlAst
+import org.partiql.lang.domains.metaContainerOf
+import org.partiql.pig.runtime.SymbolPrimitive
+
+/**
+ * A [PartiqlAst.VisitorTransform] to replace the [PartiqlAst.SortSpec] of a [PartiqlAst.OrderBy] with a reference to
+ * a [PartiqlAst.ProjectItem]'s [PartiqlAst.Expr] if an alias is provided. Also utilizes [IsTransformedOrderByAliasMeta]
+ * to enforce idempotency (delivering the same result through multiple passes).
+ *
+ * Turns:
+ *
+ * ```SELECT a + 1 AS b FROM c ORDER BY b```
+ *
+ * Into:
+ *
+ * ```SELECT a + 1 AS b FROM c ORDER BY a + 1```
+ */
+class OrderByAliasVisitorTransform : VisitorTransformBase() {
+
+    private val projectionAliases: MutableMap<String, PartiqlAst.Expr> = mutableMapOf<String, PartiqlAst.Expr>()
+
+    /**
+     * Nests itself to ensure ORDER BYs don't have access to the same [projectionAliases]
+     */
+    override fun transformExprSelect(node: PartiqlAst.Expr.Select): PartiqlAst.Expr {
+        return OrderByAliasVisitorTransform().transformExprSelectEvaluationOrder(node)
+    }
+
+    /**
+     * Uses default transform and adds the alias to the [projectionAliases] map
+     */
+    override fun transformProjectItemProjectExpr_asAlias(node: PartiqlAst.ProjectItem.ProjectExpr): SymbolPrimitive? {
+        val transformedAlias = super.transformProjectItemProjectExpr_asAlias(node)
+        if (node.asAlias != null) { projectionAliases[node.asAlias.text] = node.expr }
+        return transformedAlias
+    }
+
+    /**
+     * Uses the [OrderByAliasSupport] class to transform any encountered IDs in ORDER BY <sortSpec> into the appropriate
+     * expression using the [projectionAliases]
+     */
+    override fun transformSortSpec_expr(node: PartiqlAst.SortSpec): PartiqlAst.Expr {
+        val newExpr = when (node.expr.metas.containsKey(IsTransformedOrderByAliasMeta.TAG)) {
+            true -> super.transformSortSpec_expr(node)
+            false -> OrderByAliasSupport(projectionAliases).transformSortSpec_expr(node)
+        }
+        return newExpr.copy(metas = newExpr.metas + metaContainerOf(IsTransformedOrderByAliasMeta.instance))
+    }
+
+    /**
+     * A [PartiqlAst.VisitorTransform] that converts any found Expr.Id's into what it is mapped to in [aliases]. Also
+     * utilizes [IsTransformedOrderByAliasMeta] to ensure idempotency of this visitor (AKA visiting twice will NOT result
+     * in a different AST)
+     */
+    class OrderByAliasSupport(val aliases: Map<String, PartiqlAst.Expr>) : VisitorTransformBase() {
+        override fun transformExprId(node: PartiqlAst.Expr.Id): PartiqlAst.Expr {
+            val transformedExpr = super.transformExprId(node)
+            return when (node.case) {
+                is PartiqlAst.CaseSensitivity.CaseSensitive -> aliases[node.name.text] ?: transformedExpr
+                else -> aliases[node.name.text.toLowerCase()] ?: aliases[node.name.text.toUpperCase()] ?: transformedExpr
+            }
+        }
+    }
```

Comment:
Would you please elaborate a little more on why do we need another class here?

Reply:
So, take the following query:
```sql
SELECT (a + b) AS b FROM << {'a': 0, 'b': 1}, {'a': 0, 'b': 1} >> ORDER BY b
```

Visitor transforms only allow specifying a single behavior per datatype. So, we use "support" classes to essentially deliver multiple types of behavior. AKA, in this scenario, we only want to replace the `id` in `ORDER BY` with whatever it is aliased to. Essentially, it helps with limiting the scope of rewriting. We might not necessarily want to rewrite ALL `id`s that have value `b`.

So, instead of replacing EVERY instance of `id` (b) with a lookup, we scope it to ORDER BY. That means the example from above gets rewritten to:

```sql
SELECT (a + b) AS b FROM << {'a': 0, 'b': 1}, {'a': 0, 'b': 1} >> ORDER BY (a + b)
```

and not:

```sql
SELECT (a + (a + b)) AS b FROM << {'a': 0, 'b': 1}, {'a': 0, 'b': 1} >> ORDER BY (a + b)
```

I'm not entirely sure if it would get rewritten exactly like this because off the top of my head IDK the order of evaluation of the visitor transforms, but you get the point. We don't want to rewrite *every* encounter of `b` -- we only want to when it is in the sortSpecs of Order By. See `transformSortSpec_expr`.